### ISSUE TITLE
Enhance WASM API and Migrate to Next.js Frontend

### ISSUE THREAD

**Issue by nktr-cp** [nktr-cp/multilayer-perceptron#45]
# 🚀 Enhance WASM API and Migrate to Next.js Frontend

## 📋 Overview

The current WASM bindings and web interface expose only a fraction of the rich neural network functionality implemented in the Rust core. This issue proposes a comprehensive enhancement plan that includes:

1. **Expanding WASM API** to expose all available features
2. **Migrating from Vanilla JS to Next.js** for a modern, scalable frontend
3. **Creating a rich, interactive ML experimentation platform**

### ✅ Available in Rust Core (Fully Implemented)
- **Optimizers**: SGD, SGD with Momentum, Adam, RMSProp, GradientDescent
- **Task Types**: `BinaryClassification`, `MultiClassification`, `Regression` 
- **Regularization**: L1, L2, Elastic Net via `RegularizationConfig`
- **Loss Functions**: BinaryCrossEntropy, CrossEntropy, MeanSquaredError
- **Comprehensive Metrics**: Accuracy, Precision, Recall, F1Score, MSE, etc.
- **Data Processing**: Standardization, normalization, train/validation splits
- **Built-in Datasets**: Boston Housing (regression), Iris (multiclass), Breast Cancer (binary)
- **Model Architecture**: Flexible layer composition with various activations

### ❌ Current WASM API Limitations (`src/adapters/presentation/wasm/bindings.rs`)
- **Optimizer**: Only SGD hardcoded (line 491: `let mut optimizer = SGD::new(...)`)
- **Task Support**: Binary classification only
- **No Regularization**: L1/L2/Elastic Net not exposed
- **Limited Metrics**: Only basic accuracy available
- **No Dataset Loading**: Built-in datasets inaccessible from web
- **Fixed Architecture**: No dynamic model composition

### ❌ Current Frontend Limitations (`www/` - Vanilla JS)
- **Technology Stack**: Plain HTML/CSS/JS with basic Chart.js
- **UI Controls**: Limited to 5 basic hyperparameters
- **Model Configuration**: Single hidden layer only
- **Data Sources**: Artificial 2D points generation only
- **Visualization**: Basic loss/accuracy line charts
- **User Experience**: Static forms, no real-time interaction

### Phase 1: Next.js Frontend Setup
- [ ] **Initialize Next.js 14+ with App Router**
  - TypeScript configuration
  - Tailwind CSS for styling
  - Component-based architecture
- [ ] **WASM Integration Setup**
  - Dynamic import strategy for WebAssembly modules
  - Client-side only rendering for WASM components
  - Proper TypeScript definitions
- [ ] **Modern Development Workflow**
  - Hot reload for rapid iteration
  - Component story development
  - Proper state management (Zustand/Context)

### Phase 2: Enhanced WASM API
- [ ] **Optimizer Selection API**
  ```rust
  #[wasm_bindgen]
  pub enum OptimizerType { SGD, SGDMomentum, Adam, RMSProp, GradientDescent }
  
  #[wasm_bindgen]
  pub struct OptimizerConfig {
      optimizer_type: OptimizerType,
      learning_rate: f64,
      // Dynamic parameters based on optimizer type
      momentum: Option<f64>,     // SGD Momentum
      beta1: Option<f64>,        // Adam
      beta2: Option<f64>,        // Adam  
      epsilon: Option<f64>,      // Adam, RMSProp
  }
  ```

- [ ] **Task Type Support**
  ```rust
  #[wasm_bindgen]
  pub struct TrainingRequest {
      task_kind: TaskKind,
      optimizer_config: OptimizerConfig,
      regularization: Option<RegularizationConfig>,
      model_architecture: Vec<LayerConfig>,
      // ... other configs
  }
  ```

- [ ] **Regularization Configuration**
  ```rust
  #[wasm_bindgen]
  pub struct JsRegularizationConfig {
      l1_lambda: f64,
      l2_lambda: f64, 
      apply_to_bias: bool,
  }
  ```

- [ ] **Built-in Dataset Access**
  ```rust
  #[wasm_bindgen]
  pub fn load_boston_housing() -> JsResult<JsDataset>;
  #[wasm_bindgen] 
  pub fn load_iris_dataset() -> JsResult<JsDataset>;
  #[wasm_bindgen]
  pub fn load_breast_cancer_dataset() -> JsResult<JsDataset>;
  ```

- [ ] **Comprehensive Metrics API**
  ```rust
  #[wasm_bindgen]
  pub fn compute_classification_metrics(predictions: &JsTensor, targets: &JsTensor) -> JsClassificationMetrics;
  #[wasm_bindgen]
  pub fn compute_regression_metrics(predictions: &JsTensor, targets: &JsTensor) -> JsRegressionMetrics;
  ```

#### 🎛️ Optimizer Configuration Panel
```tsx
<OptimizerSelector 
  selected={optimizer}
  onChange={setOptimizer}
  hyperparameters={{
    learningRate: 0.001,
    momentum: 0.9,
    beta1: 0.9,
    beta2: 0.999
  }}
/>
```

#### 🏗️ Model Architecture Builder  
```tsx
<ModelArchitectureBuilder
  layers={[
    { type: 'Dense', units: 64, activation: 'ReLU' },
    { type: 'Dense', units: 32, activation: 'ReLU' },
    { type: 'Dense', units: 1, activation: 'Linear' }
  ]}
  onLayerAdd={handleAddLayer}
  onLayerRemove={handleRemoveLayer}
/>
```

#### 📊 Interactive Dataset Explorer
```tsx
<DatasetExplorer
  availableDatasets={['Boston Housing', 'Iris', 'Breast Cancer', 'Custom Upload']}
  selected={dataset}
  onSelect={setDataset}
  previewData={dataPreview}
  taskType={taskType}
/>
```

#### 📈 Real-time Training Dashboard
```tsx
<TrainingDashboard
  metrics={trainingMetrics}
  lossHistory={lossHistory}
  validationHistory={validationHistory}
  currentEpoch={epoch}
  isTraining={isTraining}
/>
```

#### 🔧 Regularization Controls
```tsx
<RegularizationPanel
  l1Lambda={l1Lambda}
  l2Lambda={l2Lambda}
  onL1Change={setL1Lambda}
  onL2Change={setL2Lambda}
  lossVisualization={regularizedLoss}
/>
```

#### 🎯 Task-Specific Views
- **Classification**: Confusion matrix, ROC curves, decision boundaries
- **Regression**: Residual plots, prediction vs actual scatter
- **Multi-task**: Tabbed interface with appropriate visualizations

### Phase 4: Advanced Features

- [ ] **Model Export/Import**
  - Save trained models as JSON
  - Load pre-trained models
  - Model comparison interface

- [ ] **Hyperparameter Optimization**
  - Grid search interface
  - Random search
  - Bayesian optimization (future)

- [ ] **Interactive Visualizations**
  - Real-time decision boundary updates (2D classification)
  - Loss landscape visualization
  - Gradient flow visualization

- [ ] **Educational Features**
  - Step-by-step training explanation
  - Mathematical formula display
  - Interactive tutorials

### Directory Structure
```
├── frontend/                 # New Next.js application
│   ├── src/
│   │   ├── app/             # App router pages
│   │   ├── components/      # Reusable UI components
│   │   ├── hooks/          # Custom React hooks for WASM
│   │   ├── lib/            # WASM integration utilities
│   │   ├── stores/         # State management
│   │   └── types/          # TypeScript definitions
│   ├── public/             # Static assets + WASM files
│   └── next.config.js      # Next.js configuration
├── src/                     # Existing Rust code
└── pkg/                    # Generated WASM bindings
```

### WASM Integration Pattern
```typescript
// hooks/useWASM.ts
export const useWASM = () => {
  const [wasmModule, setWasmModule] = useState<typeof import('../pkg/multilayer_perceptron') | null>(null);
  
  useEffect(() => {
    import('../pkg/multilayer_perceptron').then(setWasmModule);
  }, []);
  
  return wasmModule;
};
```

### Build Configuration
```javascript
// next.config.js
module.exports = {
  output: 'export',  // Static site generation
  trailingSlash: true,
  images: { unoptimized: true },
  webpack: (config) => {
    config.experiments = { ...config.experiments, asyncWebAssembly: true };
    return config;
  }
};
```

### Functional Requirements
- [ ] All 5 optimizers selectable with appropriate hyperparameters
- [ ] Support for all 3 task types (binary/multiclass/regression)
- [ ] L1/L2/Elastic Net regularization with real-time effect visualization
- [ ] All built-in datasets loadable and explorable
- [ ] Dynamic model architecture with 1-5 hidden layers
- [ ] Comprehensive metrics appropriate to each task type

### User Experience Requirements  
- [ ] Modern, responsive design with smooth interactions
- [ ] Real-time parameter updates with immediate visual feedback
- [ ] Progressive disclosure of advanced features
- [ ] Educational tooltips and explanations
- [ ] Mobile-friendly responsive layout

### Technical Requirements
- [ ] Static site generation for GitHub Pages deployment
- [ ] Bundle size optimization (target: <2MB total)
- [ ] TypeScript type safety throughout
- [ ] Component reusability and maintainability
- [ ] Cross-browser compatibility (Chrome, Firefox, Safari, Edge)

### Development Workflow
1. **WASM Build**: `wasm-pack build --target web --out-dir pkg`  
2. **Frontend Dev**: `cd frontend && npm run dev`
3. **Production Build**: `npm run build && npm run export`
4. **GitHub Pages**: Deploy `frontend/out/` directory

### CI/CD Pipeline Enhancement
```yaml
# .github/workflows/deploy.yml (updated)
- name: Build WASM
  run: wasm-pack build --target web --out-dir pkg --release
  
- name: Build Next.js  
  run: cd frontend && npm ci && npm run build
  
- name: Deploy to GitHub Pages
  uses: peaceiris/actions-gh-pages@v3
  with:
    github_token: ${{ secrets.GITHUB_TOKEN }}
    publish_dir: ./frontend/out
```

### Progressive Enhancement
- Start with basic functionality, layered with advanced features  
- Clear visual hierarchy guiding users from simple to complex workflows
- Graceful degradation for less capable browsers

### Educational Focus
- Make ML concepts accessible through interactive visualization
- Provide contextual help and mathematical explanations
- Showcase the impact of different choices in real-time

### Performance First
- Lazy loading of heavy components
- Optimized WASM bundle splitting
- Efficient React rendering with proper memoization

### Week 1-2: Foundation
- Next.js setup with TypeScript + Tailwind
- Basic WASM integration and type definitions  
- Core component structure

### Week 3-4: WASM API Expansion
- Enhanced optimizer, task, and regularization APIs
- Built-in dataset loading capabilities
- Comprehensive metrics computation

### Week 5-6: Core UI Components
- Optimizer selection and configuration
- Model architecture builder
- Dataset explorer and task selection

### Week 7-8: Advanced Features
- Real-time training dashboard
- Interactive visualizations
- Export/import functionality

### Week 9-10: Polish & Deploy
- Performance optimization
- Cross-browser testing
- Documentation and tutorials
- Production deployment

This enhancement will transform the current proof-of-concept into a comprehensive, educational neural network experimentation platform that showcases the full capabilities of the Rust implementation through a modern, accessible web interface.
